segmentstore/server/src/test/java/io/pravega/segmentstore/server/logs/ContainerMetadataUpdateTransactionTests.java

```diff
@@ -347,6 +352,102 @@ private void testWithAttributes(Function<Collection<AttributeUpdate>, Operation>
+
+        val txn = createUpdateTransaction(metadata);
+
+        // Update #1 (Value-by-reference).
+        Collection<AttributeUpdate> attributeUpdates = Arrays.asList(
+                new AttributeUpdateByReference(
+                        attributeAccumulate,
+                        AttributeUpdateType.Accumulate,
+                        new AttributeReference<>(referenceAttributeId, v -> v + 2)), // PrevValue + AttributeReference + 2
+                new AttributeUpdateByReference(
+                        attributeReplaceIfEquals,
+                        AttributeUpdateType.ReplaceIfEquals,
+                        new AttributeReference<>(referenceAttributeId, v -> v + 3), initialAttributeValue)); // AttributeReference + 3
+
+        Map<UUID, Long> expectedValues = ImmutableMap.of(
+                referenceAttributeId, initialAttributeValue,
+                attributeAccumulate, initialAttributeValue + initialAttributeValue + 2,
+                attributeNoUpdate, 123L,
+                attributeReplaceIfEquals, initialAttributeValue + 3);
+
+        Operation op = createOperation.apply(attributeUpdates);
+        txn.preProcessOperation(op);
+        txn.acceptOperation(op);
+
+        // Verify result.
+        verifyAttributeUpdates("after acceptOperation (1)", txn, attributeUpdates, expectedValues);
+
+        // Update #2 (Value-by-reference).
+        attributeUpdates = Arrays.asList(
+                new AttributeUpdateByReference(
+                        attributeAccumulate,
+                        AttributeUpdateType.Accumulate,
+                        new AttributeReference<>(attributeNoUpdate, v -> v + 20)), // PrevValue + Attr[SegmentLength] + 20
+                new AttributeUpdateByReference(
+                        attributeReplaceIfEquals,
+                        AttributeUpdateType.ReplaceIfEquals,
+                        new AttributeReference<>(attributeNoUpdate, v -> v + 30), expectedValues.get(attributeReplaceIfEquals))); // Attr[NoUpdate] + 30
+
+        expectedValues = ImmutableMap.of(
+                referenceAttributeId, initialAttributeValue,
+                attributeNoUpdate, 123L,
+                attributeAccumulate, expectedValues.get(attributeAccumulate) + expectedValues.get(attributeNoUpdate) + 20,
+                attributeReplaceIfEquals, expectedValues.get(attributeNoUpdate) + 30L);
+
+        op = createOperation.apply(attributeUpdates);
+        txn.preProcessOperation(op);
+        txn.acceptOperation(op);
+
+        // Verify result.
+        verifyAttributeUpdates("after acceptOperation (2)", txn, attributeUpdates, expectedValues);
+
+        // Update 3 (Id-by-reference).
+        // We create two new Attributes based on the value of other attributes. One is updated in the current transaction,
+        // while another is not.
+        attributeUpdates = Arrays.asList(
+                new AttributeUpdateByReference(
```

Comment:
This constructor seems to be allowing two transformations via `AttributeReference`. This doesn't seem to be strictly wrong, but to make it clear how we expect it to be used, it might be a good idea to constrain the behavior. For example, we could separate `AttributeReference` from `AttributeTransformation`. The reference is currently allowing a transformation, which enables this behavior.

Let me know if there is any misunderstanding on my side, please.